execution/gethexec/sequencer.go

```diff
@@ -1348,35 +1353,42 @@ func (s *Sequencer) updateExpectedSurplus(ctx context.Context) (int64, error) {
 	if err != nil {
 		return 0, fmt.Errorf("error encountered getting latest header from l1reader while updating expectedSurplus: %w", err)
 	}
-	l1GasPrice := header.BaseFee.Uint64()
-	if header.BlobGasUsed != nil && !s.config().Dangerous.DisableBlobBaseFeeCheck {
-		if header.ExcessBlobGas != nil {
-			blobFeePerByte, err := s.l1Reader.Client().BlobBaseFee(ctx)
-			if err != nil {
-				return 0, fmt.Errorf("error encountered getting blob base fee while updating expectedSurplus: %w", err)
-			}
-			blobFeePerByte.Mul(blobFeePerByte, blobTxBlobGasPerBlob)
-			blobFeePerByte.Div(blobFeePerByte, usableBytesInBlob)
-			if l1GasPrice > blobFeePerByte.Uint64()/16 {
-				l1GasPrice = blobFeePerByte.Uint64() / 16
-			}
+	l1GasPrice := header.BaseFee.Int64()
+
+	// #nosec G115
+	backlogCallDataUnits := int64(s.execEngine.backlogCallDataUnits())
+	var backlogCost int64 // tx's cached calldata units are already scaled by TxDataNonZeroGasEIP2028 = 16, so we divide them by 16 while calculating cost for blobs and for EIP7623 pricing accordingly
+	switch s.config().ExpectedSurplusGasPriceMode {
+	case "CalldataPrice":
+		backlogCost = backlogCallDataUnits * header.BaseFee.Int64()
+	case "BlobPrice":
+		if header.BlobGasUsed == nil || header.ExcessBlobGas == nil {
+			return 0, errors.New("expected surplus calculation is set to use blob price but latest parent chain header has BlobGasUsed or ExcessBlobGas as nil")
 		}
+		blobFeePerByte, err := s.l1Reader.Client().BlobBaseFee(ctx)
+		if err != nil {
+			return 0, fmt.Errorf("error encountered getting blob base fee while updating expectedSurplus: %w", err)
+		}
+		blobFeePerByte.Mul(blobFeePerByte, blobTxBlobGasPerBlob)
+		blobFeePerByte.Div(blobFeePerByte, usableBytesInBlob)
+		l1GasPrice = blobFeePerByte.Int64() / 16
+		backlogCost = (backlogCallDataUnits * blobFeePerByte.Int64()) / 16
+	case "CalldataPrice7623":
+		l1GasPrice = (header.BaseFee.Int64() * 40) / 16
+		backlogCost = (backlogCallDataUnits * header.BaseFee.Int64() * 40) / 16
+	default:
+		return 0, fmt.Errorf("unrecognized ExpectedSurplusGasPriceMode: %s", s.config().ExpectedSurplusGasPriceMode)
 	}
+
 	surplus, err := s.execEngine.getL1PricingSurplus()
 	if err != nil {
 		return 0, fmt.Errorf("error encountered getting l1 pricing surplus while updating expectedSurplus: %w", err)
 	}
-	// #nosec G115
-	backlogL1GasCharged := int64(s.execEngine.backlogL1GasCharged())
-	// #nosec G115
-	backlogCallDataUnits := int64(s.execEngine.backlogCallDataUnits())
-	// #nosec G115
-	expectedSurplus := int64(surplus) + backlogL1GasCharged - backlogCallDataUnits*int64(l1GasPrice)
+	expectedSurplus := surplus - backlogCost
```

Comment:
don't we still need the original expectedSurplus formula here?
```
expectedSurplus := surplus + backlogL1GasCharged - backlogCost
```
where `backlogL1GasCharged - backlogCost` is the increase/decrease of the surplus?
we charged users for `backlogL1GasCharged`, but we will need to pay `backlogCost`?

Or is the `backlogL1GasCharged` already included in surplus returned from `getL1PricingSurplus`?

Reply:
I re-read the linear ticket and it seems that `backlogL1GasCharged` already included in surplus, but I can't find it in the code. 

If the `backlogL1GasCharged` is the same as `fundsDueForRewards` from BatchPosterTable, then it is subtracted from the surplus returned in `GetL1PricingSurplus`: https://github.com/OffchainLabs/nitro/blob/1247684507fd681c89b391aa1005b1bbe8258e4e/arbos/l1pricing/l1pricing.go#L209-L210

I might be mixing up things here.